### PR TITLE
syncer: optimize safemode mechanism to improve performance

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
Implement https://github.com/pingcap/dm/issues/1757 method a.

### What is changed and how it works?
Reuse `exit_safe_binlog_name`,`exit_safe_binlog_pos`,`exit_safe_binlog_gtid` in checkpoint.
1. When parse events whose location is newer than `exit_safe_binlog`, flush `exit_safe_binlog` asap to avoid duplicate in the future.
2. Only when `syncer` fails due to errors except execution errors or meet some errors when executing, don't update global checkpoint but update safe point to downstream. In other situations, update both global checkpoint and safe point.
3. Set `syncDML` and `syncDDL`'s context to `context.Background()` to make sure ddl/dmls can be replicated to downstream as much as possible event when context is canceled.

### Check List 

Tests 

 - Unit test
 - Integration test
 
Code changes

 - Has exported function/method change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -1295,7 +1286,10 @@ func (s *Syncer) syncDML(
 			t := v.(int)
 			time.Sleep(time.Duration(t) * time.Second)
 		})
-		return db.ExecuteSQL(tctx, queries, args...)
+		// use background context to execute sqls as much as possible
+		ctctx, cancel := tctx.WithContext(context.Background()).WithTimeout(maxDMLConnectionDuration)
```
Comment (GMHDBJD) [inline syncer/syncer.go:1290]: Seems can use tctx.WithTimeout(maxDMLConnectionDuration)

Comment (lichunzhu) [reply syncer/syncer.go:1290]: Want to make sure this context is not affected by the parent context.

File: syncer/syncer.go [hunk 2]
```diff
@@ -1069,7 +1069,7 @@ func (s *Syncer) flushCheckPoints() error {
 	// optimistic shard info, DM-master may resolved the optimistic lock and let other worker execute DDL. So after this
 	// worker resume, it can not execute the DML/DDL in old binlog because of downstream table structure mismatching.
 	// We should find a way to (compensating) implement a transaction containing interaction with both etcd and SQL.
-	if err != nil {
+	if err != nil && (terror.ErrDBExecuteFailed.Equal(err) || terror.ErrDBUnExpect.Equal(err)) {
```
Comment (GMHDBJD) [inline syncer/syncer.go:1072]: Maybe only flushSafeModeExitPoint in we meet error?

**Review by GMHDBJD — COMMENTED** [review]
rest LGTM

File: syncer/syncer.go (hunk 1 above)
Comment (glorv) [reply syncer/syncer.go:1290]: I think this context with an timeout durtion of `5m` is not acceptable. In the worst case, the worker need to wait 5m before exit.

**Review by glorv — COMMENTED** [review]
rest LGTM

File: syncer/syncer.go (hunk 2 above)
Comment (lichunzhu) [reply syncer/syncer.go:1072]: For etcd error I think it's still okay to flush global checkpoint. Because it's not saved.

Comment (GMHDBJD) [reply syncer/syncer.go:1072]: I'm afraid there will be some other error that leads to data loss.

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Review by glorv — APPROVED** [review]

**Pull request merged**